<h1> Case Study Submission </h1>
<h4> Course: 040669, Simulation 1</h4>
<h4> Date: 25.11.2024 </h4>
<h4> Name: Jakob Lindstrøm </h4>
<h1> (Recommend) Online view: <a href="https://github.com/DataJakob/040669-Simulations-1-and-2/blob/main/Case_Study_Submission/deliverable.ipynb"> Submission@github</a> </h1>
<br>
<h2> i) Solution description </h2>
<p> I have solved case 2: Doctor's office. The answers to the subquestions of the case will be solved using programming languages. The answers will be written in this notebook, but some of the answers might be derived using the following repo: <a href="https://github.com/DataJakob/040669-Simulations-1-and-2/tree/main/RD_que_system"> DataJakob@github</a>.</p>
<br>

<h2> i) Reproduce the solution </h2>
<p> In IDE terminal -> git clone repo </p>
<p> Navigate to -> deliverable.ipynbd </p>
<p> Choose interpreter -> Julia  </p>
<p> Press ->  Run all </p>
<br>

<h2> iii)  General Solution Comments </h2>
<p> For this task I have choosen to use both an analytical and numerical approach to retrieve the key parameters for the first subtask.  </p>

In [9]:
include("../Formulas/MM.jl")
include("../RD_que_system/src/00_arriver.jl")
include("../RD_que_system/src/01_server.jl")
include("../RD_que_system/src/02_run_queue.jl")
include("../RD_que_system/src/03_calculate_parameters.jl")

Calculate_key_parameters (generic function with 1 method)

<h1> Task 2a </h1>
<p> To solve this task I calculate the length and time in system using both an analytical and numerical approach. </p>
<h3> "Painless Patient" </h3>
<p> Monte Carlo simulation with 100 simulations for 100 working hours. </p>

In [10]:
n_lines = 1               # number of lines
n_time = 1000.0          # 8h * 5d * 4.2w *12m
n_arr_hour = 4.0          # arrivals per hour
n_ser_hour = 5.0          # servie per hour
n_servers = 1             # Number of servers
n_packages_queue_overload = 10000   # When to switch queue line

wait_cum_MM1 = []
length_cum_MM1 =  []
for i in 1:1000
    randomness = i

    a,b,c = calculate_ia_and_s(n_lines, n_time, n_arr_hour, n_ser_hour, randomness) # arr | cum_arr | ser
    my_server = ServerBase(n_servers, n_packages_queue_overload)
    
    Data_to_json(a, c, b)
    Run_the_queue(my_server, 100000.0)
    wait_time, queue_length = Calculate_key_parameters(Int(n_time*n_arr_hour))
    push!(wait_cum_MM1, wait_time[1])
    push!(length_cum_MM1, queue_length[1])
end


MM1 = queueKeyParameters(n_arr_hour, n_ser_hour, 0)
MM1 = calcuclate_MM1(MM1)

queueKeyParameters(4.0, 5.0, 0, 0.8, 0.19999999999999996, 0.19999999999999996, 4.0, 3.2, 1.0, 0.8)

In [11]:
println("Average time in system for MM1: ", MM1.W)
println("Average length in system for MM1: ", MM1.L)
println("\n\tCompare parameters for Lq and Wq using analytical and simulation approach")
println("Analytical:")
println("Mean wait time: ",MM1.Wq)
println("Mean queue length: ",MM1.Lq)
println("\nSimulation:")
println("Mean wait time: ",round(mean(wait_cum_MM1), digits=2))
println("Mean queue length: ",round(mean(length_cum_MM1), digits=2))

Average time in system for MM1: 1.0
Average length in system for MM1: 4.0

	Compare parameters for Lq and Wq using analytical and simulation approach
Analytical:
Mean wait time: 0.8
Mean queue length: 3.2

Simulation:
Mean wait time: 0.8
Mean queue length: 3.2


<h3> "Healthy in No Time" </h3>
<p> Monte Carlo simulation with 100 simulations for 100 working hours. </p>

In [12]:
n_lines = 1               # number of lines
n_time = 1000.0          # 8h * 5d * 4.2w *12m
n_arr_hour = 4.0          # arrivals per hour
n_ser_hour = 2.5          # servie per hour
n_servers = 2             # Number of servers
n_packages_queue_overload = 1   # When to switch queue line

wait_cum_MMS = []
length_cum_MMS =  []
for i in 1:1000
    randomness = i

    a,b,c = calculate_ia_and_s(n_lines, n_time, n_arr_hour, n_ser_hour, randomness) # arr | cum_arr | ser
    my_server = ServerBase(n_servers, n_packages_queue_overload)
    
    Data_to_json(a, c, b)
    Run_the_queue(my_server, 100000.0)
    wait_time, queue_length = Calculate_key_parameters(Int(n_time*n_arr_hour))
    push!(wait_cum_MMS, wait_time[1])
    push!(length_cum_MMS, queue_length[1])
end


MMS = queueKeyParameters(n_arr_hour, n_ser_hour, 0)
MMS = calculate_MMS(MMS, n_servers)

queueKeyParameters(4.0, 2.5, 0, 0.8, 0.11111111111111109, 0.11111111111111109, 4.444444444444446, 2.844444444444446, 1.1111111111111116, 0.7111111111111115)

In [13]:
println("Average time in system for MMS: ", round(MMS.W,digits=2))
println("Average length in system for MMS: ", round(MMS.L, digits=2))
println("\n\tCompare parameters for Lq and Wq using analytical and simulation approach")
println("Analytical:")
println("Mean wait time: ",round(MMS.Wq,digits=2))
println("Mean queue length: ",round(MMS.Lq,digits=2))
println("\nSimulation:")
println("Mean wait time: ",round(mean(wait_cum_MMS), digits=2))
println("Mean queue length: ",round(mean(length_cum_MMS), digits=2))

Average time in system for MMS: 1.11
Average length in system for MMS: 4.44

	Compare parameters for Lq and Wq using analytical and simulation approach
Analytical:
Mean wait time: 0.71
Mean queue length: 2.84

Simulation:
Mean wait time: 0.68
Mean queue length: 2.71


<b> Comments: </b>
<p> The values for Lq and Wq for both alternatives calculated with both the analytical and numerical approach are close. When performing Monte Carlo with more simulations and longer times, then the values converges to the solution of the analytical methods. 
<p> Both of the systems can handle a maximum of an average 5 people per hour. However, the systems have different setups and therefore different performance.  The MM2 system that can handle two customers simultaneously and has a higher queue lengths and waiting times in the system compared to the MM1 system. </p>

<p> I believe this difference is a result of more served time in the queue system. The servers for the MM2 are twice as slow compared to the MM1 server, but there are twice as many servers in contrast. </p>
<br>

<h2> Task 2b </h2>
<p> In order to make the recommendation for Caroline's new administrative system I use the cost and the key parameters from the respective queues. The cost for the MM1 system is 14.8, and the cost for the MM2 system is 13.2.</p>
<p> I would recommend for Caroline to choose the alternative "Healthy in No Time"; the MM2 system. The reason for this is that the system has lower costs per hour, and the average length and average waiting time in the system is bigger than the other system. The recommendation is based on waiting time and queue lenght in system and the hourly costs of the systems. </p>

<h2> Task 2c - i </h2>
<p> Given the new information about the fine issued from the health insurance of the city, I need to calculate the costs for having patitents in line. </p>
<p> In order to solve this task I would calculate the amount of minutes each customer must wait during an hour multiply that with the average number of customers in a line and then multiplied that with the costs of having customers in line. 

In [14]:
println("Cost for MM2: ",round(MMS.Wq*60*MMS.Lq*3.45)+ 13.2, " per hour.")

Cost for MM2: 432.2 per hour.


<p> The hourly cost of the new adminsitrative system, given the administrative cost and the fines are 419+13.2 = 432.2. </p>
<br>
<h3> 2c - ii, Alternative recommendation </h3>
<p> Based on these results I would do the calculations for the second alternative and might reevaluate my  recommendation. Another recommendation I could propose to Caroline is to implement a finite queue. This would probably reduce her probability of having max a full queue at all times, but it would probably significantly lower her costs for having customers in queue. </p>

<p> For comparison I also calculate the costs of utilizing the MM1 system </p>

In [15]:
println("Cost for MM1: ",round(MM1.Wq*60*MM1.Lq*3.45)+14.8)

Cost for MM1: 544.8
